File: Leetcode/dfs_recursion_backtracking/279_PerfectSquares/Solution.java
class Solution {
    public int numSquares(int n) {
        int[] res = {Integer.MAX_VALUE};
        dfs(n, res, (int)Math.sqrt(n), 0);
        return res[0];
    }
    private void dfs(int target, int[] res, int num, int count){
        if(target<0 || count>=res[0])
            return;
        if(target==0){
            res[0] = Math.min(res[0], count);
            return;
        }
        for(int i = num; i>0; --i){
            int square = (int)Math.pow(i, 2);
            if( square>target)
                continue;
            dfs(target-square, res, i, count+1);
        }
    }
}

File: Leetcode/dfs_recursion_backtracking/279_PerfectSquares/Solution.py
class Solution(object):
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        res = [sys.maxsize]
        def dfs(target, num, count):
            if target<0 or count>=res[0]:
                return
            if target==0:
                res[0] = min(res[0], count)
                return
            for i in range(num, -1, -1):
                square = i**2
                if square > target:
                    continue
                dfs(target-square, i, count+1)
        dfs(n, int(n**0.5), 0)
        return res[0]
